document-portal/file-transfer.c

```diff
@@ -0,0 +1,544 @@
+
+  transfer = file_transfer_start (app_info, sender, writable, autostop);
+
+  g_dbus_method_invocation_return_value (invocation, g_variant_new ("(s)", transfer->key));
+}
+
+static void
+add_files (GDBusMethodInvocation *invocation,
+           GVariant *parameters,
+           XdpAppInfo *app_info)
+{
+  FileTransfer *transfer;
+  const char *key;
+  g_autoptr(GVariant) options = NULL;
+  GDBusMessage *message;
+  GUnixFDList *fd_list;
+  g_autoptr(GVariantIter) iter = NULL;
+  int fd_id;
+  const int *fds;
+  int n_fds;
+
+  g_variant_get (parameters, "(&sah@a{sv})", &key, &iter, &options);
+
+  transfer = lookup_transfer (key);
+  if (transfer == NULL)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid transfer");
+      return;
+    }
+
+  TRANSFER_AUTOLOCK_UNREF (transfer);
+
+  if (strcmp (transfer->sender, g_dbus_method_invocation_get_sender (invocation)) != 0)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid transfer");
+      return;
+    }
+
+  message = g_dbus_method_invocation_get_message (invocation);
+  fd_list = g_dbus_message_get_unix_fd_list (message);
+
+  fds = g_unix_fd_list_peek_fds (fd_list, &n_fds);
+
+  g_debug ("add %d files to file transfer owned by '%s' (%s)", n_fds,
+           xdp_app_info_get_id (transfer->app_info),
+           transfer->sender);
+
+  while (g_variant_iter_next (iter, "h", &fd_id))
+    {
+      int fd = -1;
+      g_autofree char *path = NULL;
+      gboolean fd_is_writable;
+      g_autofree char *dirname = NULL;
+      xdp_autofd int dir_fd = -1;
+      struct stat parent_st_buf;
+
+      if (fd_id < n_fds)
+        fd = fds[fd_id];
+
+      if (fd == -1)
+        {
+          g_dbus_method_invocation_return_error (invocation,
+                                                 G_DBUS_ERROR,
+                                                 G_DBUS_ERROR_ACCESS_DENIED,
+                                                 "Invalid transfer");
+          return;
+        }
+
+      path = xdp_app_info_get_path_for_fd (app_info, fd, 0, NULL, &fd_is_writable);
+      if (path == NULL || (transfer->writable && !fd_is_writable))
+        {
+          g_dbus_method_invocation_return_error (invocation,
+                                                 XDG_DESKTOP_PORTAL_ERROR,
+                                                 XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                                                 "Can't export file");
+          return;
+        }
+
+      dirname = g_path_get_dirname (path);
+      dir_fd = open (dirname, O_CLOEXEC | O_PATH);
```

Comment:
Additionally, there is a tiny race here too, where the app could do the attach between getting the path for the passed in fd, and the open+fstate of the parent dir to find the parent inode. In document-portal.c::validate_fd() this is done by opening the parent, verifying that you can fstatat(basename) from the dir fd and find the same inode as the fstat() on the fd that the app passed.

In fact, if you just reuse a call to that here you can drop a bunch of code and fix this race in one go.